cdap-gateway/src/main/java/io/cdap/cdap/gateway/router/handlers/AuthenticationHandler.java

```diff
@@ -108,43 +107,57 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
       ctx.fireChannelRead(msg);
       return;
     }
-    TokenState tokenState = validateAccessToken(request, ctx.channel());
-    if (tokenState.isValid()) {
-      ctx.fireChannelRead(msg);
-      return;
-    }
-
-    // Response with failure, plus optionally audit log
+    UserIdentityPair userIdentityPair;
     try {
-      HttpHeaders headers = new DefaultHttpHeaders();
-      JsonObject jsonObject = new JsonObject();
-      if (tokenState == TokenState.MISSING) {
-        headers.add(HttpHeaderNames.WWW_AUTHENTICATE, String.format("Bearer realm=\"%s\"", realm));
-        LOG.debug("Authentication failed due to missing token");
-      } else {
-        headers.add(HttpHeaderNames.WWW_AUTHENTICATE,
-                    String.format("Bearer realm=\"%s\" error=\"invalid_token\"" +
-                                    " error_description=\"%s\"", realm, tokenState.getMsg()));
-        jsonObject.addProperty("error", "invalid_token");
-        jsonObject.addProperty("error_description", tokenState.getMsg());
-        LOG.debug("Authentication failed due to invalid token, reason={};", tokenState);
+      userIdentityPair = userIdentityExtractor.extract(request);
+      // User identity extraction succeeded, so set some header properties and allow the call through
+      request.headers().remove(HttpHeaderNames.AUTHORIZATION);
+      String credential = userIdentityPair.getUserCredential();
+      // For backwards compatibility, we continue propagating credentials by default. This may change in the future.
+      if (cConf.getBoolean(Constants.Security.Authentication.AUTH_PROPAGATE_USER_CREDENTIAL, true) &&
+        credential != null) {
+        request.headers().set(HttpHeaderNames.AUTHORIZATION, "CDAP-verified " + credential);
+      }
+      request.headers().set(Constants.Security.Headers.USER_ID,
+                            userIdentityPair.getUserIdentity().getUsername());
+      String clientIP = Networks.getIP(ctx.channel().remoteAddress());
+      if (clientIP != null) {
+        request.headers().set(Constants.Security.Headers.USER_IP, clientIP);
       }
+      ctx.fireChannelRead(msg);
+    } catch (IdentityExtractionException e) {
+      // Response with failure, plus optionally audit log
+      try {
+        HttpHeaders headers = new DefaultHttpHeaders();
+        JsonObject jsonObject = new JsonObject();
+        if (e.getError().equals(IdentityExtractionException.Classification.ERROR_MISSING_CREDENTIAL)) {
```

Comment:
Generally flow control using exceptions is considered an antipattern. Also it's not very cheap to throw an exception each time when we are not in the proxy mode

Reply:
Done.